### PR TITLE
Add genericity on rustdoc generated pages

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @alexcrichton 
cc @Gankro 

This PR adds the possibility to switch style: [generated doc link](http://guillaumegomez.github.io/rust/std/index.html)

image

I generate a file to include new css files directly before compiling `librustdoc`. However, a few questions remain:
- Should I create a file which includes "basics" and only work on changes?
- Is my change on Makefile file correct?

This is a first step before adding this change to rustbook as well.

### CODE REVIEW

**Comment by Gankro** [comment]
The screenshot is busted for me, idk github seems to be puking serving the file (I once got a blurred copy, now I get a blank one).

Is there a motivation for this addition? It's kinda cool, but we shouldn't necessarily add stuff just because it's cool.

**Comment by GuillaumeGomez** [comment]
Because this white documentation is really annoying and killing my eyes. But I didn't want to impose my preferences so I added the possibility to switch between the two styles.
@steveklabnik gave me his approval to do it so I did. :)

**Comment by alexcrichton** [comment]
I'm also a little wary on adding multiple styles to documentation, is this common of other documentation generators? Perhaps the generation step could have a style, but the *rendered step*?

**Comment by GuillaumeGomez** [comment]
I find it better this way but I'm totally open to other proposition of course.

**Comment by GuillaumeGomez** [comment]
I also updated the screenshot link. And improved the code a bit.

**Comment by steveklabnik** [comment]
So, I really dislike dark themes, but it's a very, very common feature request, so I did mention to @GuillaumeGomez that they should try to work this up, to see what it's like.

Usually, what people ask for is "how do I get a dark style".

**Comment by alexcrichton** [comment]
If we're just trying to solve the questions of how to get a different style at all, aren't that what browser extensions are for? This should largely just boil down to a different stylesheet in theory.

I'd feel a little better about this if any other documentation generators had this sort of option, but I think I'd still prefer to remain consistent among all our docs. cc @brson, you've had thoughts about this in the past as well.

**Comment by GuillaumeGomez** [comment]
Asking user to do it themselves by adding browser extensions is a bit annoying. Why not providing it ourselves if we can ? It doesn't take that much after all. :-/

**Comment by lwandrebeck** [comment]
Running Gnome 3 and apps in dark theme, I must admit being able to easily get doc in dark would save my eyes a bit :+1: 
@alexcrichton you do speak about browser extension, I’d need to check but I’m not that sure that Gnome Web (Epiphany) does support that. Thus, an in-doc dark (and others ?) theme would perfectly fit (at least my) some needs.

**Comment by GuillaumeGomez** [comment]
I'd also like to precise that I want to implement it differently on rustbook : instead (or alongside, I really like the possibility to switch theme on the fly) of providing the possibility to switch the style while reading, I'd want to add the possibility to load a custom theme at book generation time.

I hope it was clear enough...

**Comment by nikomatsakis** [comment]
@alexcrichton I don't have a strong opinion here, but I think loading custom CSS is (imo) kind of a tall ask for ordinary humans.

**Comment by alexcrichton** [comment]
I do kinda agree that saying "you should use a browser extension" is quite a bit to ask, but on the other hand it also seems just as rare to request not only a different style from rustdoc but supporting multiple styles?

**Comment by GuillaumeGomez** [comment]
@alexcrichton: It's not rare for people to ask theme that fit them better, but it's rare that their request is granted, and that's a shame. I doesn't request much to do after all.

**Comment by alexcrichton** [comment]
Unfortunately an addition like this ends up not being small when viewed over the vector of time. Although this is not much to add now, this may have consequences like:

* This creates an inconsistent experience when browsing Rust documentation. If your preference is an alternate theme then you have to set that for every rustdoc instance you view. Not only that, but most Rust "properties" have the same style (e.g. the home page and the std api docs), and you won't be able to change the theme of all of them.
* Bug reports will be filed and we'll have to start asking which style everyone's using. This is just an added layer of friction which isn't always necessary.
* Bug reports will specifically be filed against one theme or another requiring us to either collect them or actively fix problems in *both* styles.
* New features in rustdoc will have to be design for all styles, not just one. It's already basically a herculean task keeping rustdoc up to date, much less keeping the style itself up to date.

**Comment by GuillaumeGomez** [comment]
I don't agree with you:

* It doesn't create an inconsistent experience but an alternative one. If it became the main theme, it'd be inconsistent but it's not the case.
* I don't see how style (where only colors in css changed) can have bugs. And if bugs there are, they're also in the main theme.
* Just like the previous point, if a bug is found on the dark theme, it certainly does exist for the main one.
* This is an issue I thought about. But here, we're talking about modifying css, and nothing more. You make a change on the main theme, you report it on the dark one and it won't take much more time (but it'll take time, yes). For this point, you can just ping me to make the change and I'll gladly do it if it allows users to have a dark theme.

**Comment by vberger** [comment]
Among the possibilities to reduce the burden of maintaining such a split, as long as the only change is the colors, wouldn't it be a possibility to split the CSS file in two ?

A file defining the structure of the page, and 2 small files for the colors (light.css / dark.css) ?

**Comment by GuillaumeGomez** [comment]
@vberger: Your explanation is way better than my:

> Should I create a file which includes "basics" and only work on changes?

File: mk/main.mk [hunk 1]
```diff
@@ -598,7 +598,7 @@ ALL_TARGET_RULES = $(foreach target,$(CFG_TARGET), \
 	$(foreach host,$(CFG_HOST), \
  all-target-$(target)-host-$(host)))
 
-all: $(ALL_TARGET_RULES) $(GENERATED) docs
+all: $(shell $(S)src/etc/generate_styles.py) $(ALL_TARGET_RULES) $(GENERATED) docs
```
Comment (wthrowe) [inline mk/main.mk]: Target dependencies are evaluated at parse time, so this is going to cause `generate_styles.py` to be run every time make is invoked.  It should be moved to a recipe somewhere.

Comment (GuillaumeGomez) [inline mk/main.mk]: I put it there in order to make it work. I couldn't figure out where was the \*right\* place but if you know it, please tell me the secret ! :)

Comment (wthrowe) [inline mk/main.mk]: I believe `$(S)src/librustdoc/html/styles.rs` should have its own target, much like [how `llvmdeps.rs` used to be handled](https://github.com/rust-lang/rust/blob/2264049577e1d64a2a863a9b9f2b7d94f6b90e15/mk/llvm.mk#L60).  The appropriate stamp file (`stamp.rustdoc`, I think) [should depend on it](https://github.com/rust-lang/rust/blob/2264049577e1d64a2a863a9b9f2b7d94f6b90e15/mk/target.mk#L137).

I'm not sure if generating the file in `src` is appropriate or if it should be somewhere in the target output directory.  If the latter it should probably look like the [current `llvmdeps.rs` code](https://github.com/rust-lang/rust/blob/master/mk/llvm.mk#L82).  If it does end up in `src` it should be presumably be added to `.gitignore`.

**Comment by brson** [comment]
I'm definitely on board with being able to customize the generated style.

I too am a little wary of being able to switch styles at runtime. As someone who likes a nice dark background though, I also understand the appeal. (As an aside, I still really want the rustonomicon to be dark by default.) I always want to resist the urge to add more options to distract people with.

Does this patch do both and can it be split into two prs? Since adding the dark theme to the live docs is going to require a lot of consideration anyway, I think it's reasonable to split this up as possible.

**Comment by GuillaumeGomez** [comment]
@brson: Like explained by @vberger, I'll create a common file with basics of rustdoc style and then just leave theme changes in both main.css and dark.css. So this PR will mainly add the possibility to switch theme. I'll provide a live rustdoc demo as soon as possible.

Thanks for your support ! :)

**Comment by GuillaumeGomez** [comment]
I splitted the css files and added `styles.rs` to gitignore file. I'll create the online rustdoc tomorrow.

**Comment by GuillaumeGomez** [comment]
And [here](http://guillaumegomez.github.io/rust/std/index.html) is the generated std doc.

**Comment by GuillaumeGomez** [comment]
I also added local storage to keep the style between pages.

**Comment by mdinger** [comment]
To those who are opposed to a dark theme for maintainability or stylistic reasons, that isn't necessarily a good reason. Some people may prefer dark themes for stylistic reason but the more important subset may use them to avoid headaches and severe eyestrain. I would never print white text on black paper for reading but I avoid all rust documentation if possible when the background is white.

Let me rephrase, if all the main rust developers couldn't sleep at night and were getting headaches because bright monitors (maybe set to minimum brightness) and white pages made life more difficult, this would probably be a very high priority issue.

Likely most people aren't affected that severely but those who find them slightly uncomfortable may really really want such a feature.

**Comment by mdinger** [comment]
You'll need to update the [deprecated](http://guillaumegomez.github.io/rust/std/primitive.f32.html#method.from_str_radix) and [unstable](http://guillaumegomez.github.io/rust/std/primitive.f32.html#method.integer_decode) colors because they don't seem to quite work with the theme.

This is pretty pleasant compared to the default. The switcher is kinda weird but so what.

**Comment by GuillaumeGomez** [comment]
@mdinger: Thanks for the feedback and your support ! I'll update it in the day.

**Comment by pnkfelix** [comment]
> more important subset may use them to avoid headaches and severe eyestrain

+1 to this.

We're not talking about having thirty different themes in the long run here, just the two (light vs dark), right?

Seems to me that having the option to switch it at run-time (rather than doc-generation time) is *crucial* for serving the purpose of helping the audience be able to read the docs in comfort.

**Comment by pnkfelix** [comment]
(just to be clear: I would definitely oppose getting rid of the light theme entirely; I am someone who likes to print documents and put them on an e-ink screen, and in that context, dark-text on light-background tends to be the way to go...)

**Comment by GuillaumeGomez** [comment]
@pnkfelix: The goal here isn't to get rid of the light theme but to give access to others alongside with the main one. Thanks for your comments !

**Comment by pnkfelix** [comment]
>  The goal here isn't to get rid of the light theme

Right, I understand that; I was just anticipating someone saying "if the light theme is causing eye strain, why don't we just get rid of it and use only the dark theme", and so I wanted to include my (selfish) reason for wanting the light theme to persist.

**Comment by GuillaumeGomez** [comment]
Well, no. It would be against my personal convictions to give access to alternatives. Getting rid of the light theme is a bad idea from my point of view, no need for you to give your reasons (and they're not that selfish because even if I prefer dark theme, I use the light one to print ^^).

**Comment by mdinger** [comment]
Not everyone gets eyestrain from computer screens. It's probably somewhat related to the specific computer screen and the individual but regardless, the same way some people strongly desire a dark theme, some strongly prefer a light theme. Those people probably don't get eyestrain enough to bother them.

That's probably why first questions like [this](https://users.rust-lang.org/t/the-little-book-of-rust-macros/3268/5) are typical (you can almost read it like: "release: nice, nice, nice, nice, can we have a light theme"). So aside from each of our selfish reasons for our specific preference (and e-paper and printing), there are probably just as many others (if not more) with the complete opposite preference as we have so removing the choice wouldn't be great.

**Comment by GuillaumeGomez** [comment]
@mdinger: I updated both items' style.

**Comment by brson** [comment]
Well, I'm conflicted now. The thing I was more convinced of (customizable static themes) is not what others want or this provides. So I guess the main question is whether we want to let people switch dynamically to a dark theme. If so, then whether the provided dark theme is suitably attractive and complementary to the light, and what this means for our other docs.

**Comment by GuillaumeGomez** [comment]
@brson: For rustbook, I have two things in mind:

* Just like rustdoc, wwe change the style dynamically.
* We specify the style when we build the book.

And... Is there other documentation ?

**Comment by brson** [comment]
@GuillaumeGomez Why does rustbook specify the style when building, and rustdoc doesn't?

Other docs include the doc index and error index, the website, and [rbe](https://github.com/rust-lang/rust-by-example) which seems to still use gitbook. None feel too important, though it would be nice to get rbe onto rustbook.

Does the way you are storing theme preferences allow one setting to be shared by all our rustbook and rustdoc projects across subdomains?

**Comment by GuillaumeGomez** [comment]
For rustbook, it's because we \*generally\* build the book for ourselves. I actually want to add both options.

For rustdoc, we impose it. And you'll notice that the change is very little for those who don't want the dark theme.

For other docs, I guess providing an alternative them, just like for rustdoc seems to be the best solution.

**Comment by mdinger** [comment]
I'm opposed to moving rbe to rustbook until multiple themes land on rustbook and frankly, [mdbook](https://github.com/azerupi/mdBook) seems to get much more effort put into it than rustbook and I'd probably port to that first. It already has the theme support.

Aside from that, there are a few things that prevent porting rbe (there was another but I can't remember what it is. Search probably wouldn't be a blocker but losing it wouldn't be good):

* cargo needs to expose what's in the Cargo.toml via some api. This would allow the structure of the examples to be used to generate the structure of the TOC (which is basically what it does already without cargo)
* a plugin or something needs to be setup to allow the embedded playpen. Hopefully it'll be better than the current one because mucking around with the output as rbe does makes errors much more difficult to interpret than in the playpen.
* we lose search by moving away from gitbook currently. Search is pretty useful.

**Comment by brson** [comment]
> For rustbook, it's because we *generally* build the book for ourselves

@GuillaumeGomez This still isn't clear to me. Ourselves meaning 'for personal use' as opposed to 'for publication'?

> mdbook seems to get much more effort put into it than rustbook and I'd probably port to that first. It already has the theme support.

@mdinger Hm, it does look better-maintained. If we moved the books out of the repo we could move them all to mdbook. We should be consolidating our tools long-term and I don't want to be on 2 (or 3!) *books.

**Comment by Gankro** [comment]
@brson I think the key insight here is that rustbook is generally invoked "directly" (although ultimately automated away by build scripts), but rustdoc is generally invoked indirectly (through `cargo doc` or similar).

**Comment by mdinger** [comment]
@brson I'm all for consolidation but for rbe's needs, I don't think the tooling is ready. Other books like the [book](http://doc.rust-lang.org/nightly/book/) and [the rustonomicon](http://doc.rust-lang.org/nightly/book/) probably don't have such strict requirements. Having the native tools external to rust should be a [boon](https://botbot.me/mozilla/rust-docs/2015-10-27/?msg=52885564&page=1) for usability too (that was day 2 of him trying to build it. Here's [day 1](https://botbot.me/mozilla/rust-docs/2015-10-26/?msg=52783616&page=1)).

**Comment by GuillaumeGomez** [comment]
I changed the way the style is changed while the DOM content is loaded. The flash shouldn't happen anymore.

**Comment by mdinger** [comment]
Deprecated and unstable look better. I'm sensing kinda a tron theme ha.

**Comment by GuillaumeGomez** [comment]
A tron theme ? Not flashy enough. And the blue should dominate more. So definitely not tron (too bad, I like tron ^^).

**Comment by brson** [comment]
Although this is well done, and there are strong advocates for it, I am going to close it without merging. The main reason is that forces us to consider two visual designs for all Rust documentation, which is presently created through 4 or 5 different tools. Furthermore, even if we were to follow through with overhauling all existing documentation in light of this concern, we will have to deal with the dual-design precedent when we overhaul and consolidate our visual designs in the future.

Thanks @GuillaumeGomez.

**Comment by mdinger** [comment]
In that case some kind of placeholder bug which details the problems blocking this from merging should be detailed. Then progress can be made towards this in the future, at the very least.

**Comment by GuillaumeGomez** [comment]
I was really hoping to finally have the official docs with dark style. Like said on IRC, I don't mind adding this dark theme everywhere and maintain it if needed. But this light theme is really uncomfortable.

**Comment by brson** [comment]
@mdinger I didn't mean to imply that there is a path forward for adding dark themes. It is something though we should reconsider in the future when doing a concerted website redesign.

**Comment by liigo** [comment]
Not force, just a choice

**Comment by mdinger** [comment]
@brson Oh. Thanks for clarifying. I understood it to mean that the reason this isn't accepted is because the implementation is via 4-5 different tools and so it needs to be consolidated. Not because light and dark isn't part of the current scheme.

**Comment by PersonneHasard1995** [comment]
According to me, to have this dark style is really a good thing. Let me explain we all have differents style, when i saw this new dark style i was a really pleasure for my eyes because the white style are hurting each time i'm seing my screen even with the minimal brightness and it distorts my reading because my eyes become tired. With the dark style it would be a really pleasure to read the documentation.

**Comment by najamelan** [comment]
I know that this is a bit old, but if I understand it right, we have to do [this](https://blog.guillaume-gomez.fr/articles/2016-09-16+Generating+doc+with+rustdoc+and+a+custom+theme) currently. It's not very ergonomic to have to find that page, pull out that cargo command every time we generate docs (and find that the css is outdated and should be modified further, and needs to be put either in every crate or we need to have the command take an absolute path). 

I would still like to see something like `cargo doc --open --dark`  supported out of the box. 

I read Alex worries about extra work and maintenance, but  it could be quite simple. Have a text file which maps 3 fields (colorname:light:dark) and have an automated build step which will use regex replace to plug the right colors in your themes. Other improvements can be made but are optional, eg. separate color related styles from layout styles, or use scss to actually have variable names for the colors, and have a dark and a light list of variable names in 2 files. 
Another option is to generate both anyways and let the user change it on the page with a button like duckduckgo does it. Where far not as good as having stuff automatically dark all the time, we might be reading documentation for quite a long time, which makes clicking a button once per opening the docs a fair improvement over the current state of affairs.
thanks for all the great work from the rust team though, I understand it's work.

**Comment by GuillaumeGomez** [comment]
Too complicated. I told it a few times already but I'll add alternative themes "soon". You'll be able to switch themes directly. At first I'll just put dark as an extra one and we'll see later on.
